napari/_vispy/visuals/volume.py

```diff
@@ -171,6 +95,127 @@
+        G[i] = next - prev + (next - curr) * 2.0 * detectAdjacentBackground(prev, next);
+    }
+
+    return G;
+}
+"""
+
+SMOOTH_GRADIENT_DEFINITION = """
+vec3 calculateGradient(vec3 loc, vec3 step) {
+    // calculate gradient within the volume by finite differences
+    // using a 3D sobel-feldman convolution kernel
+
+    // the kernel here is a 3x3 cube, centered on the sample at `loc`
+    // the kernel for G.z looks like this:
+
+    // [ +1 +2 +1 ]
+    // [ +2 +4 +2 ]    <-- "loc - step.z" is in the center
+    // [ +1 +2 +1 ]
+
+    // [  0  0  0 ]
+    // [  0  0  0 ]    <-- "loc" is in the center
+    // [  0  0  0 ]
+
+    // [ -1 -2 -1 ]
+    // [ -2 -4 -2 ]    <-- "loc + step.z" is in the center
+    // [ -1 -2 -1 ]
+
+    // kernels for G.x and G.y similar, but transposed
+    // see https://en.wikipedia.org/wiki/Sobel_operator#Extension_to_other_dimensions
+
+    vec3 G = vec3(0.0);
+    float curr = 0.0;
+    // next and prev are the directly adjacent values along x, y, and z
+    vec3 next = vec3(0.0);
+    vec3 prev = vec3(0.0);
+
+    float val;
+    bool is_on_border = false;
+    for (int i=-1; i <= 1; i++) {
+        for (int j=-1; j <= 1; j++) {
+            for (int k=-1; k <= 1; k++) {
+                if (is_on_border && (i != 0 && j != 0 && k != 0)) {
+                    // we only care about on-axis values if we are on a border
+                    continue;
+                }
+                vec3 sample_loc = loc + vec3(i, j, k) * step;
+                bool is_in_bounds = all(greaterThanEqual(sample_loc, vec3(0.0)))
+                    && all(lessThanEqual(sample_loc, vec3(1.0)));
+
+                if (is_in_bounds || u_clamp_at_border) {
+                    val = colorToVal($get_data(sample_loc));
+                } else {
+                    val = categorical_bg_value;
+                }
+
+                G.x += val * -float(i) *
+                    (1 + float(j == 0 || k == 0) + 2 * float(j == 0 && k == 0));
+                G.y += val * -float(j) *
+                    (1 + float(i == 0 || k == 0) + 2 * float(i == 0 && k == 0));
+                G.z += val * -float(k) *
+                    (1 + float(i == 0 || j == 0) + 2 * float(i == 0 && j == 0));
+
+                curr += int(i == 0 && j == 0 && k == 0) * val;
+                next.x += int(i == 1 && j == 0 && k == 0) * val;
+                next.y += int(i == 0 && j == 1 && k == 0) * val;
+                next.z += int(i == 0 && j == 0 && k == 1) * val;
+                prev.x += int(i == -1 && j == 0 && k == 0) * val;
+                prev.y += int(i == 0 && j == -1 && k == 0) * val;
+                prev.z += int(i == 0 && j == 0 && k == -1) * val;
+
+                is_on_border = is_on_border || (!is_in_bounds && (i == 0 || j == 0 || k == 0));
+            }
+        }
+    }
+
+    if (is_on_border) {
+        // fallback to simple gradient calculation if we are on the border
+        G = next - prev;
```

Comment:
Maybe add a comment here that this is because it looks weird otherwise LOL

Also, how weird? I wonder if it's worth the extra branching.

Reply:
I will at least add to the comment here about it.

I'll also take another look. I was having an issue where it made the label almost look like it was 1px away from the edge of the volume instead of right up against it.

Reply:
Ah I just played with it and now I remember why I added it. With this included it makes the in the current mode (dark/hollow faces) more consistent between `fast` and `smooth`.  I think it looks okay in the new mode though.

What do you think of this?

```
    if (is_on_border && u_clamp_at_border) {
        // fallback to simple gradient calculation if we are on the border
        // and clamping is enabled (old behavior with dark/hollow faces at the border)
        // this makes the faces in `fast` and `smooth` look the same in both clamping modes
        G = next - prev;
    } else {

```